Port `sort` to structured kernels.

ysiraichi opened the pull request:
Tracking Issue: #55070

This PR relands #67016, with the modifications discussed in https://github.com/pytorch/pytorch/pull/67015#issuecomment-982004500.

In summary, we call `infer_dense_strides` on the input's strides, and pass it to `set_output`. Meaning that if one of the outputs is resized (by a `resize_output` call), we will also restride such an output using the dense strides of the input.

ysiraichi committed:
Port and create tests for non-contiguous input.

ysiraichi committed:
Fix `OpInfo` tests for `sort` and `msort`.

bdhirsh commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -14,30 +14,68 @@
+      : at::infer_dense_strides(self.sizes(), self.strides());
+
+  // 'set_output' will only modify the strides, if the output was resized.
+  // Which means that we can ignore the output's strides here.
```
is this comment out of date? (strides aren't ignored)

bdhirsh replied:
ohh "ignore the output's strides here" is probably referring to whatever `maybe_get_output()[0].strides()` currently is, when we're in the out= case.

nvm, i think that comment is phrased fine.

ysiraichi replied:
Nah. This comment wasn't very well worded. I will change it explaining the `infer_dense_strides` call.

bdhirsh commented on aten/src/ATen/native/cuda/Sort.cpp:
```diff
@@ -60,37 +63,9 @@ std::tuple<Tensor &,Tensor &> sort_out_stable_cuda(const Tensor & self, c10::opt
   TORCH_CHECK(self_dtype != ScalarType::ComplexFloat && self_dtype != ScalarType::ComplexDouble,
```
welp, more cuda-only checks. There probably isn't much we can do about these though... so I think it's fine to leave them here.

bdhirsh commented on aten/src/ATen/native/cuda/Sort.cpp:
```diff
@@ -161,20 +111,11 @@ std::tuple<Tensor &,Tensor &> sort_out_stable_cuda(const Tensor & self, c10::opt
+// TODO: we should handle this accordingly when we start using REGISTER_HIP_DISPATCH,
+// since REGISTER_DISPATCH won't work in this cpp file.
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+REGISTER_CUDA_DISPATCH(sort_stub, &sort_cuda_kernel);
```
where did this come from?

ysiraichi replied:
I don't think I got what you mean with this question. I needed this, since I am using `DispatchStub`, here.

bdhirsh replied:
Just looked at this a little more. Is using `DispatchStub` for the cuda kernel actually necessary? My understanding of `DispatchStub` is:
- it allows you to do dynamic dispatch (technically redundant, since the dispatcher lets you do this too)
- it re-compiles the cpu kernel with multiple vector instruction sets

So the `sort` kernel is already using `DispatchStub` in the cpu kernel (which results in having to do 2 dynamic dispatches - one to go to CPU, and a second to go to the correct CPU vector instruction set).

But it looks like the cuda kernel in master today doesn't bother with dispatch stub. We use the dispatcher to dynamically dispatch to the cuda kernel and that's it. If the `REGISTER_CUDA_DISPATCH` is important for some other functionality (like hip support), then we can probably save that for a future PR, since it will also result in a second dynamic dispatch in the cuda path.

Lmk what you think!

ysiraichi replied:
It's true that I didn't really need `DispatchStub`. That said, I think it's nice here since we get to unify common `CPU` and `CUDA` implementation, in exchange of a little overhead (I think). Do you think it's better to duplicate the code?

> it re-compiles the cpu kernel with multiple vector instruction sets

I thought that every source file inside `native/cpu` got compiled with different instructions sets (not really dispatched to them, though). Am I missing something?

bdhirsh replied:
Ok, I think it's pretty reasonable (you're right, and we make that same tradeoff in a bunch of other ops that are implemented with DispatchStub). Sounds good!

> I thought that every source file inside native/cpu got compiled with different instructions sets (not really dispatched to them, though). Am I missing something?

Yep, that's my understanding too. And then the `sort_stub()` call is technically a dynamic dispatch that will pick the "correct" pre-compiled vectorized cpu kernel (in addition to potentially picking the cuda kernel), based on whatever hardware you happening to be using when running the op.

bdhirsh commented on test/test_sort_and_select.py:
```diff
@@ -167,6 +167,28 @@ def test_sort_large(self, device, dtype):
+        tensor = inputs[0][:, 0]
+        # Outputs: non-contiguous (stride: 5) 2-element array
+        values = inputs[1][0:2, 0]
+        indices = inputs[2][0:2, 0]
```
Hmmm.. will this test the same behavior from the original [bug](https://github.com/pytorch/pytorch/issues/65863)? I think so but I'm not sure.

In the original repro example, the two output tensors were actually too small, vs. here they are views into a larger tensor of the right size. Might be worth just making this test more similar to the repro?

ysiraichi replied:
You are right. I did it like this because I wanted to make it explicit that the strides weren't `(1,)` before. But, I guess that's not really necessary. Will change it.

bdhirsh commented:
Just left some minor comments but looks good! Thanks for working on re-landing this!

ysiraichi commented on aten/src/ATen/native/cuda/Sort.cpp:
```diff
@@ -39,15 +40,17 @@ void fillSliceWithIndex(Tensor& t,int dim) {
   // this algorithm is always stable
-  TORCH_INTERNAL_ASSERT(stable.has_value(), "sort_out(): c10::optional<bool> for stable has to have value.");
-  TensorArg self_arg{self, "self", 1}, values_arg{values, "values", 2}, indices_arg{indices, "indices", 3};
-  checkAllSameGPU(__func__, {self_arg, values_arg, indices_arg});
```
- `TORCH_INTERNAL_ASSERT(stable.has_value(), ...)`: already being checked on `META`
- `checkAllSameGPU(...)`: already checked via codegen

ysiraichi commented on aten/src/ATen/native/Sorting.cpp:
```diff
@@ -14,30 +14,68 @@
+  set_output(0, self.sizes(), strides, self.options(), {});
+  set_output(1, self.sizes(), strides, self.options().dtype(kLong), {});
+
+  // Overlap Checking
+  // 1. 'self' can be the same as 'values' (in-place).
+  // 2. 'self' can't overlap with 'indices'
+  // 3. 'values' and 'indices' can't overlap.
+  if (values_defined) {
+    at::assert_no_internal_overlap(values);
+    at::assert_no_partial_overlap(self, values);
   }
+
+  if (indices_defined) {
+    at::assert_no_internal_overlap(indices);
+    at::assert_no_overlap(self, indices);
+  }
+
+  if (values_defined && indices_defined) {
+    at::assert_no_overlap(values, indices);
+  }
```
Do you think these overlap checks make sense? The `sort` kernel didn't really check for them before, but I thought I would introduce, anyway.

bdhirsh replied:
oh, thanks for calling that out.

That seems like a pretty reasonable check, since otherwise you'd get silently incorrect results.

I would still make it a separate PR though, that way there aren't any unnecessarily "functionality changes" in the port itself. Do you mind making it a separate PR after this one in the stack

ysiraichi committed:
Improve comment about the use of strides.

ysiraichi committed:
Make the test closer to the one in the issue.

bdhirsh reviewed (approved):
LGTM! (XLA test failure looks flaky so I kicked it off again).

bdhirsh commented on aten/src/ATen/native/cuda/Sort.cpp:
```diff
@@ -39,15 +40,17 @@ void fillSliceWithIndex(Tensor& t,int dim) {
-  checkAllSameGPU(__func__, {self_arg, values_arg, indices_arg});
 
   bool is_non_overlapping_and_dense = self.is_non_overlapping_and_dense();
   int64_t ndim = self.dim();
```
internal tests are complaining about the unused variable here

ysiraichi committed:
Remove overlapping checks.

ysiraichi committed:
Purge unused variable.

ysiraichi committed:
Fix some compiling issues.

ysiraichi committed:
Move `copy` before `DispatchStub`.

This allows us to use `TensorBase` instead of `Tensor`. Thus, reducing compiling overhead.

ysiraichi committed:
Use `Tensor` (instead of `TensorBase`) on `sort` stub signature.

ezyang commented:
just waitin for it to build :>

ysiraichi committed:
Revert "Use `Tensor` (instead of `TensorBase`) on `sort` stub signature."

This reverts commit e1192defbcbc93157c60ffc19728857e1fe0dadb.

ysiraichi committed:
Convert `TensorBase` into `Tensor`, instead.

ysiraichi committed:
Fix `OpInfo` test skips for `sort` and `msort`.

ysiraichi committed:
Fix `TensorBase` -> `Tensor` conversion.

Lifetime extension wasn't being used, since we were trying to access a member
variable (`TensorImpl::target_`) indirectly (`intrusive_ptr::operator*`).

Couldn't avoid the "long way": save the intermediary `OptionalTensorRef` in a temporary
variable.

ysiraichi commented:
@ezyang I think the tests are finally passing. There is a failure, though: `Meta Internal-Only Changes Check`. Should I do something about it?

seemethere commented:
@pytorchbot revert this, broke upstream builds for [linux-xenial-cuda11.3-py3.7-gcc7-no-ops / build](https://github.com/pytorch/pytorch/runs/6081113248?check_suite_focus=true#logs)

janeyx99 commented:
Adding the label ciflow/trunk should run that failing build in your reland PR!

ezyang commented:
```
/var/lib/jenkins/workspace/aten/src/ATen/native/cuda/TensorTopK.cpp: In member function ‘void at::native::structured_topk_out_cuda::impl(const at::Tensor&, int64_t, int64_t, bool, bool, const at::Tensor&, const at::Tensor&)’:
/var/lib/jenkins/workspace/aten/src/ATen/native/cuda/TensorTopK.cpp:86:17: error: ‘sort_outf’ is not a member of ‘at::cuda’
       at::cuda::sort_outf(values, /* stable= */ false, dim, largest, sortedValues, sortedIndices);
                 ^~~~~~~~~
```

prolly missing a header include

ysiraichi committed:
Add include.

ezyang commented:
@pytorchbot merge this
